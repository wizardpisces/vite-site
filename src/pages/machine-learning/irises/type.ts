export type SampleType = {
    input: number[];
    expectedType: number;
    gotType: number;
}