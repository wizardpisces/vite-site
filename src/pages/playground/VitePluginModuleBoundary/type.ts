export type EnvsType = Record<string, boolean>