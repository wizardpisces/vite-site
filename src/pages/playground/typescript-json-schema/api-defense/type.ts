export type Response<T> = {
    data: T
    code: number
}

export enum ValidateAndFixAPIError {
    TypeMismatch = 'TypeMismatch',
    Required = 'Required',
    SchemaContainsRecursiveReference = 'SchemaContainsRecursiveReference',
    Other = 'Other',
}

export type OnErrorOptions = {
    msg: string,
    type?: ValidateAndFixAPIError,
    schema?: Record<string, any>,
    data?: Record<string, any>,
    errorDetail?: any
}

export type SchemaType = 'string' | 'number' | Record<string, any>
export type FullSchemaType = Record<string, any>