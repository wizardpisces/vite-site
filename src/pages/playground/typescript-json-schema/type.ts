export type Response<T> = {
    data: T
    code: number
}