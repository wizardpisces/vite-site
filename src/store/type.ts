export enum Mutations {
    TOGGLE_TRANSLATE_KEY='TOGGLE_TRANSLATE_KEY'
}