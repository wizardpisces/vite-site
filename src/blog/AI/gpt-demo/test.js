// prompting 生成 fn 函数满足以下测试用例：

function test0(fn) {
    if (fn(0) === 0 && fn(1) === 1) {
        return true
    }
    return false
}

function test1(fn) {
    if (fn(0) === 0 && fn(1) === 1 && fn(2) === 1 && fn(3) === 2) {
        return true
    }
    return false
}