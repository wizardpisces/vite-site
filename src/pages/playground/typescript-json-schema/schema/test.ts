// type Recursive = {
//     recur: Recursive
// } | number

// export type ApiSchema = {
//     testArray: Recursive
// }   
// export type ApiSchema = {
//     testArray: [],
//     testObject: {
//         arr: [],
//         obj: {
//             name: string,
//             age: number,
//             recursive: Recursive
//         },
//         optionalObj?: {
//             key1: string,
//             key2: number
//         }
//         simpleType: string
//     },
// }

export type ApiSchema = {
    name: string,
    age: number,
    otherApi: ApiSchema2
}

export type ApiSchema2 = {
    name: string,
    age: number
}