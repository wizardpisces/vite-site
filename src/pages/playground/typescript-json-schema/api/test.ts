import scheme from './scheme.json'
import { ApiSchema } from "../schema/test";
import { validateAPI } from "../utils";
import { Response } from '../type';

const schema = scheme['definitions']['ApiSchema']

export const getTestData = ()=> {
        const res: Response<ApiSchema> = {
            code:0,
            data:{
                testArray: [],
                testChainReference: null,
                testReference:{
                    basicInfo: {
                        name: 'test',
                        age: 18
                    },
                    otherInfo: {
                        
                    }
                },
                testObject: {
                    arr: [],
                    // obj: {
                    //     name: 'test',
                    //     age: 18,
                    //     recursive:1
                    // },
                    optionalObj: {
                        key1: 'test',
                        key2: 18
                    },
                    simpleType: 'test'
                }
            }
        }
    return validateAPI(res, schema)
}

