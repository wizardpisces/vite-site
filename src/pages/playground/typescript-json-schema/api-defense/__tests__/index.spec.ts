import { beforeEach, describe, expect, jest, test } from '@jest/globals';
import { getApiDefenseFn } from '..';
import { BasicObj, Recursive, TestChainReference } from './schema/test';
import { ValidateAndFixAPIError } from '../type';
import fullSchema from './api/jsonschema/test.json'

global.console.error = () => { } // 避免 reportError 打印错误，影响到测试结果查看

const onError = jest.fn((obj) => { })
const validateAndFixAPI = getApiDefenseFn({ fullSchema, fullSchemaId: 'test', onError });

describe('validateAPI', () => {
    beforeEach(() => {
        onError.mockClear()
    })
    test('should support basic null completion, support nested object and array, will skip optional and simple type and type not defined in schema', () => {
        const data: BasicObj = {
            // @ts-ignore
            arr: null,
            // @ts-ignore
            obj: null,
            notInSchemaDefinition: null
        }
        const expected: BasicObj = {
            arr: [], // 数组补全
            obj: {
                obj2: {} // 嵌套补全
            },
            // @ts-ignore
            notInSchemaDefinition: null // schema 未定则不做处理
            // simpleType 普通类型未做补全
        }
        expect(validateAndFixAPI(data, fullSchema['definitions']['BasicObj'])).toEqual(expected)
        // expect(onError).toHaveBeenCalled()
        expect(onError).toHaveBeenCalledTimes(1)
        expect(onError.mock.calls[0]).toMatchSnapshot()
        // @ts-ignore
        expect((onError.mock.calls[0][0]).type).toEqual(ValidateAndFixAPIError.Required) // simpleType is required
    })

    test('should support $ref chain', () => {
        // @ts-ignore
        const data: TestChainReference = {
        }
        const expected: TestChainReference = {
            // @ts-ignore
            reference: { // 类型定义中引用其他类型
                // testChain:1
            }
        }
        expect(validateAndFixAPI(data, fullSchema['definitions']['TestChainReference'])).toEqual(expected)

        expect(onError).toHaveBeenCalledTimes(1)
        expect(onError.mock.calls[0]).toMatchSnapshot()
        // @ts-ignore
        expect((onError.mock.calls[0][0]).type).toEqual(ValidateAndFixAPIError.Required) // reference is required

    })

    test('should support cyclic $ref skip null completion, and report error', () => {
        // @ts-ignore
        const data: Recursive = {
        }
        // @ts-ignore
        const expected: Recursive = { // Recursive type contains circular reference, so it will not be completed
        }

        expect(validateAndFixAPI(data, fullSchema['definitions']['Recursive'])).toEqual(expected)

        expect(onError.mock.calls).toMatchSnapshot()
        expect(onError).toHaveBeenCalledTimes(2)
        // @ts-ignore
        expect((onError.mock.calls[0][0]).type).toEqual(ValidateAndFixAPIError.Required) // recur is required
        // @ts-ignore
        expect((onError.mock.calls[1][0]).type).toEqual(ValidateAndFixAPIError.SchemaContainsRecursiveReference) //  SchemaContainsRecursiveReference
    })
});