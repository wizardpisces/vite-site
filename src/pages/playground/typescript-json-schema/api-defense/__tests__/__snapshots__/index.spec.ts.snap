// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateAPI should support $ref chain 1`] = `
[
  {
    "data": {
      "reference": {},
    },
    "errorDetail": {
      "instancePath": "",
      "keyword": "required",
      "message": "must have required property 'reference'",
      "params": {
        "missingProperty": "reference",
      },
      "schemaPath": "#/required",
    },
    "msg": "missing property",
    "schema": {
      "additionalProperties": false,
      "properties": {
        "reference": {
          "$ref": "test#/definitions/ChainReference",
        },
      },
      "required": [
        "reference",
      ],
      "type": "object",
    },
    "type": "Required",
  },
]
`;

exports[`validateAPI should support basic null completion, support nested object and array, will skip optional and simple type and type not defined in schema 1`] = `
[
  {
    "data": {
      "arr": [],
      "notInSchemaDefinition": null,
      "obj": {
        "obj2": {},
      },
    },
    "errorDetail": {
      "instancePath": "",
      "keyword": "required",
      "message": "must have required property 'simpleType'",
      "params": {
        "missingProperty": "simpleType",
      },
      "schemaPath": "#/required",
    },
    "msg": "missing property",
    "schema": {
      "additionalProperties": false,
      "properties": {
        "arr": {
          "maxItems": 0,
          "minItems": 0,
          "type": "array",
        },
        "obj": {
          "additionalProperties": false,
          "properties": {
            "obj2": {
              "additionalProperties": false,
              "type": "object",
            },
          },
          "required": [
            "obj2",
          ],
          "type": "object",
        },
        "optionalObj": {
          "additionalProperties": false,
          "properties": {
            "key1": {
              "type": "string",
            },
            "key2": {
              "type": "number",
            },
          },
          "required": [
            "key1",
            "key2",
          ],
          "type": "object",
        },
        "simpleType": {
          "type": "string",
        },
      },
      "required": [
        "arr",
        "obj",
        "simpleType",
      ],
      "type": "object",
    },
    "type": "Required",
  },
]
`;

exports[`validateAPI should support cyclic $ref skip null completion, and report error 1`] = `
[
  [
    {
      "data": {},
      "errorDetail": {
        "instancePath": "",
        "keyword": "required",
        "message": "must have required property 'recur'",
        "params": {
          "missingProperty": "recur",
        },
        "schemaPath": "#/anyOf/0/required",
      },
      "msg": "missing property",
      "schema": {
        "anyOf": [
          {
            "additionalProperties": false,
            "properties": {
              "recur": {
                "$ref": "test#/definitions/Recursive",
              },
            },
            "required": [
              "recur",
            ],
            "type": "object",
          },
          {
            "type": "number",
          },
        ],
      },
      "type": "Required",
    },
  ],
  [
    {
      "data": {},
      "msg": "Schema contains cyclic reference, will not do complete work; please check test#/definitions/Recursive or use Nullable to fix potential error",
      "schema": {
        "anyOf": [
          {
            "additionalProperties": false,
            "properties": {
              "recur": {
                "$ref": "test#/definitions/Recursive",
              },
            },
            "required": [
              "recur",
            ],
            "type": "object",
          },
          {
            "type": "number",
          },
        ],
      },
      "type": "SchemaContainsRecursiveReference",
    },
  ],
]
`;
